filestore.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// Read satisfies io.Reader
+func (f *File) Read(p []byte) (n int, err error) {
+	if f.closed {
+		return 0, errors.New("file is closed")
+	}
+	if !f.readOnly {
+		return 0, errors.New("file is not open for reads")
+	}
+
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	bytesRemaining := f.arraySize - f.bytesRead
+	if bytesRemaining == 0 {
+		return 0, io.EOF
+	}
+	if int64(len(p)) > bytesRemaining {
+		p = p[0:bytesRemaining]
+	}
+
+	cArrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(cArrayURI))
+	if err := bufferExport(f.tdbCtx, cArrayURI, f.bytesRead, p); err != nil {
+		return 0, err
+	}
+	f.bytesRead += int64(len(p))
+
+	return len(p), nil
+}
+
+// Write satisfies io.Writer
+func (f *File) Write(p []byte) (n int, err error) {
+	if f.closed {
+		return 0, errors.New("file is closed")
+	}
+	if f.readOnly {
+		return 0, errors.New("file is not open for writes")
+	}
+	return f.buf.Write(p)
+}
+
+// Close satisfies io.Closer
+func (f *File) Close() error {
+	if f.closed {
+		return nil
+	}
+
+	defer func() {
+		f.closed = true
+	}()
+
+	if f.createSchema {
+		schema, err := NewArraySchemaForFile(f.tdbCtx, "")
+		if err != nil {
+			return err
+		}
+		array, err := NewArray(f.tdbCtx, f.arrayURI)
+		if err != nil {
+			return err
+		}
+		err = array.Create(schema)
+		if err != nil {
+			return err
+		}
+	}
+
+	cArrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(cArrayURI))
+	return bufferImport(f.tdbCtx, cArrayURI, f.buf.Bytes(), f.mimeType)
+}
+
+// CreateFile creates a TileDB array at arrayURI, suitable to store a file of mimeType, and opens it for writing.
+// The returned File is an inmemory writer. Each Write stores the data in an internal bytes.Buffer.
+// The actual TileDB array creation happens on Close.
+func (fs *FileStore) CreateFile(arrayURI string, mimeType FileStoreMimeType) (*File, error) {
+	return &File{
+		tdbCtx:       fs.tdbCtx,
+		arrayURI:     arrayURI,
+		createSchema: true,
+		mimeType:     mimeType,
+	}, nil
+}
+
+// OpenFileRead opens for reading the array at arrayURI, which should have a filestore schema.
+func (fs *FileStore) OpenFileToRead(arrayURI string) (*File, error) {
+	siz, err := fs.Size(arrayURI)
+	if err != nil {
+		return nil, err
+	}
+	return &File{
+		tdbCtx:    fs.tdbCtx,
+		arrayURI:  arrayURI,
+		readOnly:  true,
+		arraySize: siz,
+	}, nil
+}
+
+// OpenFileWrite opens for writing the array at arrayURI, which should have a filestore schema.
+// The returned File is an inmemory writer. Each Write stores the data in an internal bytes.Buffer.
+// The actual TileDB write happens on Close.
+func (fs *FileStore) OpenFileToWrite(arrayURI string) (*File, error) {
+	return &File{
+		tdbCtx:   fs.tdbCtx,
+		arrayURI: arrayURI,
+	}, nil
+}
+
+// NewArraySchemaForFile allocs a new ArraySchema optimized for the storage of file.
+// An empty path returns a general schema suitable for any file.
+func NewArraySchemaForFile(tdbCtx *Context, filePath string) (*ArraySchema, error) {
```

Comment:
Does this need to be exposed? Do we think users will create file array schemas without using the files api as an entrypoint?

Reply:
For completeness i think it must be exposed. The core API exposes it and all other binding we provide py,java,C# expose it also